zipline/_protocol.pyx

```diff
@@ -45,6 +45,11 @@ cdef class BarData:
         try:
             view = self._views[asset]
         except KeyError:
+            try:
+                asset = self.data_portal.env.asset_finder.retrieve_asset(asset)
```

Comment:
What this does is ensure that the key is always an Asset derived class. The inverse of our paradigm of calling `int(sid)` when passing to persistence layers.

Both an `int` and corresponding Asset class will both has to the same int value. The `KeyError` here is raised on first attempt to access any asset/sid in from the BarData, whether it is via `data[sid(24)]` or `data[24]`. We then do a `defaultdict`-like behavior and insert a `SidView` on first access. The problem was that if an int was used for the first access the key in `self._views` and asset passed to the `SidView` was an integer. Which lead to issues when trying to call data_portal methods that expected an asset.

A unit test exercising the blotter was creating an Order with an integer `sid`, which uncovered this behavior because of the new use of `BarData` in the blotter.

An alternative fix would be to ensure that both tests and production Order's all have an `Asset` assigned to the `sid` value.